### PR TITLE
Feature/metadata

### PR DESCRIPTION
**Pull request by jokedurnez**
This PR stores in a metadata entity (`ol.surveymetadata`) for each participant: 
- first datetime logged
- last datetime logged
- all dates logged

### CODE REVIEW

File: src/main/java/com/openlattice/chronicle/constants/EdmConstants.java [hunk 1]
```diff
@@ -40,11 +42,14 @@ private EdmConstants() {
     public static final FullQualifiedName MODEL_FQN       = new FullQualifiedName( "vehicle.model" );
     public static final FullQualifiedName DATE_TIME_FQN   = new FullQualifiedName( "ol.datetime" );
     public static final FullQualifiedName START_DATE_TIME = new FullQualifiedName( "ol.datetimestart" );
+    public static final FullQualifiedName END_DATE_TIME   = new FullQualifiedName( "ol.datetimeend" );
+    public static final FullQualifiedName RECORDED_DATE_TIME_FQN = new FullQualifiedName( "ol.recordeddate" );
```
Comment (anzioka) [inline src/main/java/com/openlattice/chronicle/constants/EdmConstants.java:46]: Could you run code reformat? (Code -> Reformat Code). You would need to configure formatting settings on IntelliJ though.  

**How set up on IntelliJ**

1. Save the file at https://github.com/openlattice/openlattice/blob/master/eclipse-code-format-settings.xml on your desktop
2. Open Preferences. Select code style -> Java -> Import Scheme -> Eclipse XML Profile

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 2]
```diff
@@ -110,7 +116,10 @@
     private final UUID fullNamePTID;
     private final UUID recordTypePTID;
     private final UUID startDateTimePTID;
+    private final UUID endDateTimePTID;
+    private final UUID recordedDateTimePTID;
     private final UUID durationPTID;
+    private final UUID olidPTID;
```
Comment (anzioka) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:122]: camel case?

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 3]
```diff
@@ -385,6 +424,92 @@ private void createUserAppsEntitiesAndAssociations(
...
+        dataApi.updateEntitiesInEntitySet( metadataESID,
+                ImmutableMap.of( metadataEntityKeyId, metadataEntityData ),
+                UpdateType.PartialReplace );
+
+        Map<UUID, Set<Object>> hasEntityData = new HashMap<>();
+        hasEntityData.put( olIdPTID, Set.of( firstDateTime ) );
+        UUID hasEntityKeyId = reserveHasEntityKeyId( metadataEntityData, dataIntegrationApi );
+        dataApi.updateEntitiesInEntitySet( hasESID,
+                ImmutableMap.of( hasEntityKeyId, hasEntityData ),
+                UpdateType.Merge );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:502]: Why `UpdateType.Merge`?

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 4]
```diff
@@ -385,6 +424,92 @@ private void createUserAppsEntitiesAndAssociations(
...
+                                        .parse( date.toString() )
+                                        .toInstant()
+                                        .atZone( tz )
+                                        .toOffsetDateTime() );
+
+                    }
+                    return dateTimes;
+                } )
+                .flatMap( Set::stream )
+                .filter( datetime -> datetime.isAfter( MINIMUM_DATE ) )
```
[the remark this answers is not in the document]
Comment (jokedurnez) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:461]: Yeah.  Not sure if the 1970 only shows up with usage stats, or also with other recordTypes.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java (hunk 3 above)
Comment (jokedurnez) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java]: For no reason and it's gone :)

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jokedurnez) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:470]: How would I do that?  The data first needs to be interpreted as datetime (map) to allow for filtering, and then flattened from `Stream<HashSet<OffsetDateTime>>` to `Stream<OffsetDateTime>` I think.

[the remark this answers is not in the document]
Comment (jokedurnez) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:470]: Maybe change `map` to `forEach` and add only filtered to a new Set maybe rather?

[the remark this answers is not in the document]
Comment (jokedurnez) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java]: Changed into a forEach.

**Pull request merged**